### PR TITLE
[Felix] Add retries when netlink list APIs encounter EINTR.

### PR DESCRIPTION
**Pull request by sridhartigera**
## Description

Add retries to netlink List APIs as it now returns EINTR. 

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/dataplane/linux/int_dataplane.go [hunk 1]
```diff
@@ -1106,7 +1107,14 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
 // findHostMTU auto-detects the smallest host interface MTU.
 func findHostMTU(matchRegex *regexp.Regexp) (int, error) {
 	// Find all the interfaces on the host.
-	links, err := netlink.LinkList()
+
+	nlHandle, err := netlinkshim.NewRealNetlink()
+	if err != nil {
+		log.WithError(err).Error("failed to created netlink handle. Unable to auto-detect MTU")
```
Comment (fasaxc) [inline felix/dataplane/linux/int_dataplane.go:1113]:
```suggestion
		log.WithError(err).Error("Failed to create netlink handle. Unable to auto-detect MTU.")
```

File: felix/fv/etcd_restart_test.go [hunk 2]
```diff
@@ -60,7 +58,7 @@ var _ = Context("etcd connection interruption", func() {
 		// Wait until the tunl0 device appears; it is created when felix inserts the ipip module
 		// into the kernel.
 		Eventually(func() error {
-			links, err := netlink.LinkList()
+			links, err := utils.LinkList()
```
Comment (fasaxc) [inline felix/fv/etcd_restart_test.go:61]: `utils` is best avoided, it ends up being a catchall and you could have 100 utils packages with different random stuff in them.

Can we emphasize the retry somehow; important for the user to know that this is not he same as the basic version.

`netlinkretry.LinkList()`
`netlinkutils.LinkListRetryEINTR()`

File: felix/netlinkshim/netlink.go [hunk 3]
```diff
@@ -47,6 +49,156 @@ type Interface interface {
...
 }
 
+type RealNetlink struct {
+	nlHandle *netlink.Handle
+}
+
 func NewRealNetlink() (Interface, error) {
-	return netlink.NewHandle(syscall.NETLINK_ROUTE)
+	nlHandle, err := netlink.NewHandle(syscall.NETLINK_ROUTE)
+	return &RealNetlink{
```
Comment (fasaxc) [inline felix/netlinkshim/netlink.go:58]: Prefer to return `nil, err` if there's an error.  Makes incorrect use fail quicker.

File: felix/netlinkshim/netlink.go [hunk 4]
```diff
@@ -47,6 +49,156 @@ type Interface interface {
...
+func (r *RealNetlink) RuleAdd(rule *netlink.Rule) error {
+	return r.nlHandle.RuleAdd(rule)
+}
+
+func (r *RealNetlink) RuleDel(rule *netlink.Rule) error {
+	return r.nlHandle.RuleDel(rule)
+}
+
+func (r *RealNetlink) Delete() {
+	//nolint:staticcheck
```
Comment (fasaxc) [inline felix/netlinkshim/netlink.go:176]: What's the linter's problem?

**Review by fasaxc — CHANGES_REQUESTED** [review]
Pragmatic to re-use the existing shim.  Saves proliferation.  Just need to remember that the shim is needed now and that it has function in it rather than being a trivial shim

File: felix/netlinkshim/netlink.go (hunk 4 above)
Comment (sridhartigera) [reply felix/netlinkshim/netlink.go:179]: ```netlink.Delete``` is deprecated and I got the below error from the linter. 
```SA1019: r.nlHandle.Delete is deprecated: use Close instead which is in line with typical resource release patterns for files and other resources.```

Comment (fasaxc) [reply felix/netlinkshim/netlink.go:179]: Delete just calls through to Close so let's just switch to using Close?

**Review by fasaxc — APPROVED** [review]

**Pull request merged**